### PR TITLE
Minor: add additional documentation about `BufWriter`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes #.

# Rationale for this change
Inspired by https://github.com/apache/arrow-datafusion/issues/9614, I wanted to make it clearer that BufWriter existed / what it could do

# What changes are included in this PR?
1. Update main lib docs to mention `BufReader` and `BufWriter`
2. Update docs for `BufReader` to mention it adaptively used put / put multi-part

# Are there any user-facing changes?
Docs

No code / behavior changes

### CODE REVIEW

File: object_store/src/buffered.rs [hunk 1]
```diff
@@ -207,6 +207,10 @@ impl AsyncBufRead for BufReader {
 
 /// An async buffered writer compatible with the tokio IO traits
 ///
+/// This writer adaptively uses [`ObjectStore::put`] or
```
Comment (alamb) [inline object_store/src/buffered.rs:210]: this is such an important feature I think it is worth calling out explicitly

File: object_store/src/buffered.rs [hunk 2]
```diff
@@ -255,7 +259,8 @@ impl BufWriter {
         }
     }
 
-    /// Returns the [`MultipartId`] if multipart upload
+    /// Returns the [`MultipartId`] of the last multipart upload created by this
```
Comment (alamb) [inline object_store/src/buffered.rs:262]: I didn't completely follow the code here. 

```suggestion
    /// Returns the [`MultipartId`] of multipart upload created by this
```

File: object_store/src/lib.rs [hunk 3]
```diff
@@ -88,19 +88,24 @@
...
-//! This provides some compelling advantages:
+//! This design provides the following advantages:
 //!
 //! * All operations are atomic, and readers cannot observe partial and/or failed writes
 //! * Methods map directly to object store APIs, providing both efficiency and predictability
 //! * Abstracts away filesystem and operating system specific quirks, ensuring portability
 //! * Allows for functionality not native to filesystems, such as operation preconditions
 //! and atomic multipart uploads
 //!
+//! This crate does provide a [`BufReader`] and [`BufWriter`] adapters
```
Comment (tustvold) [inline object_store/src/lib.rs:103]:
```suggestion
//! This crate does provide [`BufReader`] and [`BufWriter`] adapters
```

File: object_store/src/lib.rs [hunk 4]
```diff
@@ -88,19 +88,24 @@
...
+//! which provide a more filesystem-like API for working with the
+//! [`ObjectStore`] trait.
```
Comment (tustvold) [inline object_store/src/lib.rs:105]:
```suggestion
//! [`ObjectStore`] trait, however, they should be used with care 
```

**Review by tustvold — COMMENTED** [review]
Just a minor copy suggestion, I think we should highlight that they're a compatibility tool and not hugely recommended

**Pull request merged**